File: design-bundle/project/engine.js
// ─────────────────────────────────────────────────────────────
// DermDiag ES — Inference Engine
// Forward chaining + MYCIN-style Certainty Factor combination
// ─────────────────────────────────────────────────────────────

(function () {
  const KB = window.DERM_KB;

  // MYCIN-style CF combination for two positive CFs in [0,1]
  function combineCF(a, b) {
    if (a == null) return b;
    if (b == null) return a;
    // both positive: a + b * (1 - a)
    return a + b * (1 - a);
  }

  // Evaluate one rule against working memory (facts).
  // Returns null if rule does not fire, otherwise an object with cf + bindings.
  function evaluateRule(rule, facts) {
    const sx = facts.symptoms || {};        // set of selected symptom keys → true
    const meta = facts.meta || {};          // single-value fields

    // requires (all): every key in `requires` must be true in symptoms
    if (rule.requires) {
      for (const k of rule.requires) {
        if (!sx[k]) return null;
      }
    }

    // requiresAny: at least one key must be true
    if (rule.requiresAny) {
      const any = rule.requiresAny.some(k => sx[k]);
      if (!any) return null;
    }

    // excludes: none of the listed keys may be true
    if (rule.excludes) {
      for (const k of rule.excludes) {
        if (sx[k]) return null;
      }
    }

    // requiresMeta: each key's value must be in the allowed list
    if (rule.requiresMeta) {
      for (const [key, allowed] of Object.entries(rule.requiresMeta)) {
        const v = meta[key];
        if (v == null || !allowed.includes(v)) return null;
      }
    }

    return { cf: rule.cf, ruleId: rule.id, dx: rule.dx, label: rule.label, expr: rule.expr };
  }

  // Forward-chaining: scan rules, fire matching ones, accumulate per-diagnosis CFs.
  function runInference(facts) {
    const fired = [];
    const cfByDx = {};
    const traceByDx = {};

    for (const rule of KB.rules) {
      const result = evaluateRule(rule, facts);
      if (!result) continue;
      fired.push(result);
      cfByDx[result.dx] = combineCF(cfByDx[result.dx], result.cf);
      if (!traceByDx[result.dx]) traceByDx[result.dx] = [];
      traceByDx[result.dx].push(result);
    }

    // Build ranked diagnoses
    const diagnoses = Object.entries(cfByDx)
      .map(([dx, cf]) => ({
        dx,
        cf: Math.round(cf * 100) / 100,
        condition: KB.conditions[dx],
        firedRules: traceByDx[dx]
      }))
      .sort((a, b) => b.cf - a.cf);

    // Red flag detection (independent of rules)
    const redFlags = [];
    for (const rf of KB.redFlags) {
      if (facts.symptoms && facts.symptoms[rf.key]) {
        redFlags.push(rf);
      }
    }

    // Overall verdict
    let verdict = "low";
    if (diagnoses.length > 0) {
      const top = diagnoses[0].cf;
      if (top >= 0.85) verdict = "high";
      else if (top >= 0.6) verdict = "moderate";
      else verdict = "low";
    }

    return {
      diagnoses,
      redFlags,
      verdict,
      firedCount: fired.length,
      facts
    };
  }

  // Helper: confidence band for a single diagnosis
  function confidenceBand(cf) {
    if (cf >= 0.85) return { lvl: "high", text: "High confidence" };
    if (cf >= 0.6)  return { lvl: "med",  text: "Moderate confidence" };
    return { lvl: "low", text: "Low confidence — consult professional" };
  }

  window.DERM_ENGINE = { runInference, confidenceBand, combineCF };
})();

File: backend/internal/engine/engine.go
// Package engine implements forward-chaining inference over the DermDiag KB
// with MYCIN-style certainty-factor combination. Ported from engine.js.
package engine

import (
	"math"
	"sort"

	"github.com/krr/dermdiag/internal/kb"
)

type Facts struct {
	Symptoms map[string]bool   `json:"symptoms"`
	Meta     map[string]string `json:"meta"`
}

type FiredRule struct {
	RuleID string  `json:"ruleId"`
	Dx     string  `json:"dx"`
	CF     float64 `json:"cf"`
	Label  string  `json:"label"`
	Expr   string  `json:"expr"`
}

type Diagnosis struct {
	Dx          string       `json:"dx"`
	CF          float64      `json:"cf"`
	Condition   kb.Condition `json:"condition"`
	FiredRules  []FiredRule  `json:"firedRules"`
}

type Result struct {
	Diagnoses  []Diagnosis  `json:"diagnoses"`
	RedFlags   []kb.RedFlag `json:"redFlags"`
	Verdict    string       `json:"verdict"`
	FiredCount int          `json:"firedCount"`
	Facts      Facts        `json:"facts"`
}

func combineCF(a, b float64) float64 {
	return a + b*(1-a)
}

func contains(xs []string, x string) bool {
	for _, v := range xs {
		if v == x {
			return true
		}
	}
	return false
}

func evaluate(rule kb.Rule, f Facts) *FiredRule {
	for _, k := range rule.Requires {
		if !f.Symptoms[k] {
			return nil
		}
	}
	if len(rule.RequiresAny) > 0 {
		any := false
		for _, k := range rule.RequiresAny {
			if f.Symptoms[k] {
				any = true
				break
			}
		}
		if !any {
			return nil
		}
	}
	for _, k := range rule.Excludes {
		if f.Symptoms[k] {
			return nil
		}
	}
	for key, allowed := range rule.RequiresMeta {
		v, ok := f.Meta[key]
		if !ok || !contains(allowed, v) {
			return nil
		}
	}
	return &FiredRule{RuleID: rule.ID, Dx: rule.Dx, CF: rule.CF, Label: rule.Label, Expr: rule.Expr}
}

func Run(f Facts) Result {
	if f.Symptoms == nil {
		f.Symptoms = map[string]bool{}
	}
	if f.Meta == nil {
		f.Meta = map[string]string{}
	}

	cfByDx := map[string]float64{}
	traceByDx := map[string][]FiredRule{}
	fired := 0

	for _, rule := range kb.Rules {
		r := evaluate(rule, f)
		if r == nil {
			continue
		}
		fired++
		if cur, ok := cfByDx[r.Dx]; ok {
			cfByDx[r.Dx] = combineCF(cur, r.CF)
		} else {
			cfByDx[r.Dx] = r.CF
		}
		traceByDx[r.Dx] = append(traceByDx[r.Dx], *r)
	}

	diagnoses := make([]Diagnosis, 0, len(cfByDx))
	for dx, cf := range cfByDx {
		diagnoses = append(diagnoses, Diagnosis{
			Dx:         dx,
			CF:         math.Round(cf*100) / 100,
			Condition:  kb.Conditions[dx],
			FiredRules: traceByDx[dx],
		})
	}
	sort.Slice(diagnoses, func(i, j int) bool { return diagnoses[i].CF > diagnoses[j].CF })

	var redFlags []kb.RedFlag
	for _, rf := range kb.RedFlags {
		if f.Symptoms[rf.Key] {
			redFlags = append(redFlags, rf)
		}
	}

	verdict := "low"
	if len(diagnoses) > 0 {
		top := diagnoses[0].CF
		switch {
		case top >= 0.85:
			verdict = "high"
		case top >= 0.6:
			verdict = "moderate"
		}
	}

	return Result{
		Diagnoses:  diagnoses,
		RedFlags:   redFlags,
		Verdict:    verdict,
		FiredCount: fired,
		Facts:      f,
	}
}
